UniversalAlgebra/Subalgebras/Setoid.lagda

```diff
@@ -0,0 +1,244 @@
+  A = 𝕌[ 𝑨 ]
+
+ ⋂s : {ι : Level}(I : Type ι){ρ : Level}{𝒜 : I → Pred A ρ}
+  →   (∀ i → 𝒜 i ∈ Subuniverses 𝑨) → ⋂ I 𝒜 ∈ Subuniverses 𝑨
+
+ ⋂s I σ f a ν = λ i → σ i f a (λ x → ν x i)
+
+\end{code}
+
+In the proof above, we assume the following typing judgments:
+
+```
+ν  : Im a ⊆ ⋂ I 𝒜
+a  : ∥ 𝑆 ∥ f → Subalgebras.Setoid.A 𝑨
+f  : ∣ 𝑆 ∣
+σ  : (i : I) → 𝒜 i ∈ Subuniverses 𝑨
+```
+and we must prove `(f ̂ 𝑨) a ∈ ⋂ I 𝒜`.  When we did this with the old
+Algebra type, Agda could fill in the proof term `λ i → σ i f a (λ x → ν x i)`
+automatically using `C-c C-a`, but this doesn't work for SetoidAlgebra
+as we've implemented it.  We get the error "Agsy does not support copatterns
+yet."  We should fix the implementation to resolve this.
+
+\begin{code}
+
+module _ {χ : Level}{X : Type χ}
+         {α ρᵃ : Level}{𝑨 : SetoidAlgebra α ρᵃ}
+         where
+
+ private A = 𝕌[ 𝑨 ]
+ open Setoid
+ open Environment 𝑨
+ open Func renaming ( f to _<$>_ )
+
+ -- subuniverses are closed under the action of term operations
+ sub-term-closed : (B : Pred A ρ)
+  →                (B ∈ Subuniverses 𝑨)
+  →                (t : Term X)
+  →                (b : Carrier (Env X))
+  →                (∀ x → (b x ∈ B)) → (⟦ t ⟧ <$> b) ∈ B
+
+ sub-term-closed _ _ (ℊ x) b Bb = Bb x
+
+ sub-term-closed B B≤A (node f t)b ν =
+  B≤A f  (λ z → ⟦ t z ⟧ <$> b) λ x → sub-term-closed B B≤A (t x) b ν
+
+\end{code}
+
+In the induction step of the foregoing proof, the typing judgments of the premise are the following:
+
+```
+ν  : (x : X) → b x ∈ B
+b  : Setoid.Carrier (Env X)
+t  : ∥ 𝑆 ∥ f → Term X
+f  : ∣ 𝑆 ∣
+σ  : B ∈ Subuniverses 𝑨
+B  : Pred A ρ
+ρ  : Level
+𝑨  : SetoidAlgebra α ρᵃ
+```
+and the given proof term establishes the goal `⟦ node f t ⟧ <$> b ∈ B`.
+
+Alternatively, we could express the preceeding fact using an inductive type representing images of terms.
+
+\begin{code}
+
+ data TermImage (B : Pred A ρ) : Pred A (𝓞 ⊔ 𝓥 ⊔ α ⊔ ρ)
+  where
+  var : ∀ {b : A} → b ∈ B → b ∈ TermImage B
+  app : ∀ f ts →  ((i : ∥ 𝑆 ∥ f) → ts i ∈ TermImage B)  → (f ̂ 𝑨) ts ∈ TermImage B
+
+ -- `TermImage B` is a subuniverse of 𝑨 that contains B.
+ TermImageIsSub : {B : Pred A ρ} → TermImage B ∈ Subuniverses 𝑨
+ TermImageIsSub = app
+
+ B-onlyif-TermImageB : {B : Pred A ρ} → B ⊆ TermImage B
+ B-onlyif-TermImageB Ba = var Ba
+
+ -- Since `Sg B` is the smallest subuniverse containing B, we obtain the following inclusion.
+ SgB-onlyif-TermImageB : (B : Pred A ρ) → Sg 𝑨 B ⊆ TermImage B
+ SgB-onlyif-TermImageB B = sgIsSmallest 𝑨 (TermImage B) TermImageIsSub B-onlyif-TermImageB
+
+ module _ {β ρᵇ : Level}{𝑩 : SetoidAlgebra β ρᵇ} where
+
+  private B = 𝕌[ 𝑩 ]
+  open Environment 𝑩
+
+  -- Homomorphisms are uniquely determined by their values on a generating set.
+  hom-unique : swelldef 𝓥 β → (G : Pred A ρ)  (g h : hom 𝑨 𝑩)
+   →           ((x : A) → (x ∈ G → ∣ g ∣ x ≡ ∣ h ∣ x))
+               -------------------------------------------------
+   →           (a : A) → (a ∈ Sg 𝑨 G → ∣ g ∣ a ≡ ∣ h ∣ a)
```

Comment:
As 𝑩 is a SetoidAlgebra, you should use its equivalence here. Something here still seems odd though, I can't quite put my finger on it.

Reply:
Yes, I've been thinking about this (including in my dreams last night :)  I'm thinking we need an equivalence of tuples of elements of SetoidAlgebras.  The "right" thing to do isn't yet obvious (to me).

Reply:
I've opened a new issue about this.

I'm happy with how the other stuff worked about, especially with the new additions to Subalgebras.Setoids involving the subalgebra and isomorphism relations...  should I wait for you to look at these recent additions, or just merge?